---
title: "Other Analyses"
author: 
  - Huang Lin$^1$
  - $^1$NICHD, 6710B Rockledge Dr, Bethesda, MD 20892
date: '`r format(Sys.Date(), "%B %d, %Y")`'
output:
  html_document: 
    toc: true
    theme: united
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(echo = TRUE, warning = FALSE, 
                      message = FALSE, comment = NA,
                      fig.width = 6.25, fig.height = 5)

library(tidyverse)
library(microbiome)
library(SpiecEasi)
library(lubridate)
library(ggpubr)
library(DT)
options(DT.options = list(
  initComplete = JS("function(settings, json) {",
  "$(this.api().table().header()).css({'background-color': '#000', 'color': '#fff'});","}")))

source("../programs/00_secom.R")

get_upper_tri = function(cormat){
    cormat[lower.tri(cormat)] = NA
    diag(cormat) = NA
    return(cormat)
}

hard_thresh = function(R, th){
  R_th = R
  R_th[abs(R) <= th] = 0
  return(R_th)
}
```

# 1. Comparison of computational efficiency

```{r, eval=FALSE}
linear_data_generation = function(n, d, d1, corr_mu, corr_prob, 
                                  uncorr_mu, uncorr_prob, dispersion, seed) {
  set.seed(seed)
  d2 = d - d1
  
  mu = c(sample(corr_mu, d1, replace = TRUE, prob = corr_prob), # Correlated taxa
         sample(uncorr_mu, d2, replace = TRUE, prob = uncorr_prob)) # Uncorrelated taxa
  
  # Absolute abundances
  A = matrix(NA, ncol = n, nrow = d)
  for (i in seq_len(d)) {
    A[i, ] = rnbinom(n = n, size = 1/dispersion, mu = mu[i])
  }
  
  for (i in seq(2, d1, 2)) {
    A[i, ] = poly(x = A[i - 1, ], degree = 1, raw = FALSE)
    A[i, ] = round(mu[i] * A[i, ]) + mu[i]
  }
  
  R0 = matrix(0, ncol = d, nrow = d)
  lmat = replicate(d1/2, 
                   matrix(1, ncol = 2, nrow = 2), 
                   simplify = FALSE)
  R0_sub = as.matrix(Matrix::bdiag(lmat))
  R0_sub[R0_sub == 1] = cor(t(log(A[seq_len(d1), ] + 1)))[R0_sub == 1]
  R0[seq_len(d1), seq_len(d1)] = R0_sub
  
  # Sequencing efficiency
  C = exp(rnorm(d, mean = 0, sd = 1))
  
  # Microbial loads in the ecosystem
  A_prim = A * C
  A_dot = colSums(A_prim)
  
  # Relative abundances in the ecosystem
  R = A_prim/t(replicate(d, A_dot))
  
  # Sampling fractions
  S = rbeta(n = n, shape1 = 2, shape2 = 10)
  
  # Library sizes
  O_dot = round(S * A_dot)
  # Observed abundances
  O = matrix(NA, nrow = d, ncol = n)
  for (i in seq(n)) {
    O[, i] = rmultinom(1, size = O_dot[i], prob = R[, i])
  }
  
  res = list(O = O, R0 = R0)
  return(res)
}

n = 100
d = 200
d1 = 50
corr_mu = c(2000, 10000, 40000, 100000)
corr_prob = c(0.1, 0.4, 0.4, 0.1)
uncorr_mu = c(2000, 10000, 40000, 100000)
uncorr_prob = c(0.1, 0.4, 0.4, 0.1)
dispersion = 0.5
seed = 123

sim_data = linear_data_generation(n, d, d1, corr_mu, corr_prob, 
                                  uncorr_mu, uncorr_prob, dispersion, seed)
O = sim_data$O
R0 = sim_data$R0
taxa_id = paste0("T", seq_len(d))
sample_id = paste0("S", seq_len(n))
meta_data = data.frame(sample_id = sample_id)
dimnames(O) = list(taxa_id, sample_id)
OTU = otu_table(O, taxa_are_rows = TRUE)
META = sample_data(meta_data)
sample_names(META) = meta_data$sample_id
otu_data = phyloseq(OTU, META)

# SECOM
phyloseq1 = otu_data; phyloseq2 = otu_data
zero_cut = 0.5; lib_cut = 1000; wins_quant = c(0, 1)
method = "pearson"; soft = FALSE; composit_cor = 0.5; thresh_len = 20
n_cv = 10; seed = 123; thresh_hard = 0.3; max_p = 0.005; n_cl = 1

t_start = Sys.time()
res_secom = secom_linear(phyloseq1, phyloseq2, zero_cut, lib_cut, 
                         wins_quant, method, soft, composit_cor, 
                         thresh_len, n_cv, seed, thresh_hard, max_p, n_cl)
t_end = Sys.time()
t_secom = difftime(t_end, t_start, units = "secs")

# Proportionality
vlr = function(mat, x) {
  lr = log(mat/x)
  vlr = apply(lr, 2, var)
  return(vlr)
}

t_start = Sys.time()
Rel = sweep(O, 2, colSums(O, na.rm = TRUE), "/")
Rel_clr = as.data.frame(clr(t(Rel)))
Rel_clr_var = apply(Rel_clr, 2, var) 
Rel_vlr = apply(t(Rel), 2, function(x) vlr(t(Rel), x))

Rel_phi = sweep(Rel_vlr, 2, Rel_clr_var, FUN = "/")
R_hat_prop = matrix(NA, nrow = d, ncol = d)
for (i in seq_len(d)) {
  for (j in seq(i, d)) {
    R_hat_prop[i, j] = min(c(Rel_phi[i, j], Rel_phi[j, i]))
  }
}
t_end = Sys.time()
t_prop = difftime(t_end, t_start, units = "secs")

# SparCC
t_start = Sys.time()
R_hat_sparcc = sparcc(t(O))$Cor
t_end = Sys.time()
t_sparcc = difftime(t_end, t_start, units = "secs")

# SPIEC-EASI (MB)
t_start = Sys.time()
se_mb = spiec.easi(t(O), method = "mb", lambda.min.ratio = 1e-2,
                   nlambda = 20, pulsar.params = list(rep.num = 50))
t_end = Sys.time()
t_mb = difftime(t_end, t_start, units = "secs")

# SPIEC-EASI (GL)
t_start = Sys.time()
se_gl = spiec.easi(t(O), method = "glasso", lambda.min.ratio = 1e-2,
                   nlambda = 20, pulsar.params = list(rep.num = 50))
t_end = Sys.time()
t_gl = difftime(t_end, t_start, units = "secs")

# Standard Pearson
t_start = Sys.time()
res_sample = Hmisc::rcorr(t(O), type = "pearson")
corr_sample = res_sample$r
corr_sample_p = res_sample$P
diag(corr_sample_p) = 0
t_end = Sys.time()
t_sample = difftime(t_end, t_start, units = "secs")

# Output
df_t = data.frame(SECOM = t_secom, Proportionality = t_prop,
                  SparCC = t_sparcc, `SPIEC-EASI (MB)` = t_mb,
                  `SPIEC-EASI (GL)` = t_gl, Pearson = t_sample, 
                  check.names = FALSE)
df_t = df_t %>%
  mutate_all(function(x) round(x, 1))

write_csv(df_t, "../outputs/others/cpu_time.csv")
```

```{r}
df_t = read_csv("../outputs/others/cpu_time.csv")
datatable(df_t)
```

# 2. Linear vs. distance correlations

```{r, fig.height=5, fig.width=10}
df = data.frame(x = seq(0, 10, 0.5)) %>% 
  mutate(Linear = 3 * x,
         Quadratic = (x - 5)^2)
  
ann = data.frame(x = c(5, 5), y = c(25.5, 25.5), 
                 label = c(paste0("pCor = ", round(cor.test(df$x, df$Linear, method = "pearson")$estimate, 2),
                                  "\nsCor = ", round(cor.test(df$x, df$Linear, method = "spearman")$estimate, 2),
                                  "\ndCor = ", round(dcor(df$x, df$Linear), 2)),
                           paste0("pCor = ", round(cor.test(df$x, df$Quadratic, method = "pearson")$estimate, 2),
                                  "\nsCor = ", round(cor.test(df$x, df$Quadratic, method = "spearman")$estimate, 2),
                                  "\ndCor = ", round(dcor(df$x, df$Quadratic), 2))),
                 type = c("Linear", "Quadratic"))
df = df %>%
  pivot_longer(cols = Linear:Quadratic, names_to = "type", values_to = "y")

p = ggplot(data = df, aes(x = x, y = y)) +
  geom_point() +
  scale_y_continuous(breaks = seq(0, 25, 5), limits = c(0, 27)) + 
  facet_grid(.~type) +
  geom_label(data = ann, aes(x = x, y = y, label = label),
             color = "orange", size = 4, fontface = "bold") +
  theme_bw() +
  theme(strip.background = element_rect(fill = "white"))
p
ggsave(plot = p, "../images/supp/supp_diff_corr.pdf", height = 5, width = 10)   
ggsave(plot = p, "../images/supp/supp_diff_corr.jpeg", height = 5, width = 10, dpi = 300)
```

# 3. Subset of data vs. full data

## 3.1 Data import

```{r}
# Metadata
meta_data = read_tsv("../data/global_gut/global_gut_metadata.txt")
meta_data = meta_data %>% 
  transmute(sample_id = `#SampleID`, 
            age = as.numeric(AGE), 
            age_group = case_when(
              age <= 2 ~ "Age \u2264 2",
              age >= 18 ~ "Age \u2265 18",
              TRUE ~ "Others"
            ),
            sex = SEX, 
            country = case_when(
              COUNTRY == "GAZ:Malawi" ~ "MA",
              COUNTRY == "GAZ:United States of America" ~ "US",
              COUNTRY == "GAZ:Venezuela" ~ "VEN",
              TRUE ~ "Missing"
            )) %>%
  arrange(sample_id)
meta_data = meta_data[-nrow(meta_data), ] # The last row is non-informative
meta_data$country = factor(meta_data$country, levels = c("MA", "US", "VEN"))
meta_data$sex = factor(meta_data$sex, levels = c("female", "male"))

# Taxonomy
tax = read_tsv("../data/global_gut/global_gut_taxonomy.txt") %>% 
  arrange(OTU_ID)
otu_id = tax$OTU_ID
tax = data.frame(tax[, -1]) %>%
  rowwise() %>%
  dplyr::mutate_all(function(x) strsplit(x, "__")[[1]][2]) %>%
  mutate(Species = strsplit(Species, " ")[[1]][2],
         Species = ifelse(!is.na(Species) & !is.na(Genus),
                           paste(strsplit(Genus, "")[[1]][1], Species, sep = "."),
                           NA)) %>%
  ungroup()
tax = as.matrix(tax)
rownames(tax) = otu_id

# OTU table
otu_table = read_tsv("../data/global_gut/global_gut_otu.txt")
otu_table = otu_table[, -ncol(otu_table)]
otu_id = otu_table$OTU_ID
otu_table = data.frame(otu_table[, -1], check.names = FALSE, row.names = otu_id)
otu_table = as.matrix(otu_table)

# OTU data
OTU = otu_table(otu_table, taxa_are_rows = TRUE)
META = sample_data(meta_data)
sample_names(META) = meta_data$sample_id
TAX = tax_table(tax)
otu_data = phyloseq(OTU, TAX, META)

# Subset samples
otu_data = subset_samples(otu_data, age_group == "Age \u2265 18")

# Aggregate to species level
species_data = aggregate_taxa(otu_data, "Species")
species_data = subset_taxa(species_data, Species != "Unknown")

# Aggregate to genus level
Clostridium_ind = data.frame(tax) %>%
  rownames_to_column("otu_id") %>%
  filter(Genus == "Clostridium") %>%
  .$otu_id
Ruminococcus_ind = data.frame(tax) %>%
  rownames_to_column("otu_id") %>%
  filter(Genus == "Ruminococcus") %>%
  .$otu_id
Bacteroides_ind = data.frame(tax) %>%
  rownames_to_column("otu_id") %>%
  filter(Genus == "Bacteroides") %>%
  .$otu_id
Eubacterium_ind = data.frame(tax) %>%
  rownames_to_column("otu_id") %>%
  filter(Genus == "Eubacterium") %>%
  .$otu_id
  
otu_data2 = merge_taxa(otu_data, 
                       taxa_names(otu_data)[taxa_names(otu_data) %in% Clostridium_ind], 
                       1)
otu_data2 = merge_taxa(otu_data2, 
                       taxa_names(otu_data2)[taxa_names(otu_data2) %in% Ruminococcus_ind], 
                       1)
otu_data2 = merge_taxa(otu_data2, 
                       taxa_names(otu_data2)[taxa_names(otu_data2) %in% Bacteroides_ind], 
                       1)
otu_data2 = merge_taxa(otu_data2, 
                       taxa_names(otu_data2)[taxa_names(otu_data2) %in% Eubacterium_ind], 
                       1)
otu_data2 = phyloseq(otu_table(abundances(otu_data2), taxa_are_rows = TRUE), 
                     TAX, META)
genus_data = aggregate_taxa(otu_data2, "Genus")
genus_data = subset_taxa(genus_data, Genus != "Unknown")
```

## 3.2 Visualizations {.tabset}

### SECOM (subset of data)

```{r}
phyloseq1 = species_data
topx2 = genus_data %>%
  top_taxa(n = 10)
phyloseq2 = subset_taxa(genus_data, Genus %in% topx2)

pseqs = list(c(phyloseq2, phyloseq2))
pesudo = 0; zero_cut = 0.5; corr_cut = 0.5; lib_cut = 1000
wins_quant = c(0.05, 0.95); method = "pearson"; soft = FALSE; thresh_len = 20
n_cv = 10; seed = 123; thresh_hard = 0.3; max_p = 0.005; n_cl = 1

res_secom1 = secom_linear(pseqs, pesudo, zero_cut, corr_cut, lib_cut, 
                          wins_quant, method, soft, thresh_len, n_cv, 
                          seed, thresh_hard, max_p, n_cl)

df = res_secom1$corr_fl
df_p = data.frame(get_upper_tri(df)) %>%
  rownames_to_column("var1") %>%
  pivot_longer(cols = colnames(df), names_to = "var2", 
               values_to = "value") %>%
  filter(!is.na(value)) %>%
  mutate(value = round(value, 2))
df_p$var1 = factor(df_p$var1, levels = colnames(df))
df_p$var2 = factor(df_p$var2, levels = colnames(df))

p_secom1 = df_p %>%
  ggplot(aes(var2, var1, fill = value)) +
  geom_tile(color = "black") +
  scale_fill_gradient2(low = "blue", high = "red", mid = "white", na.value = "white",
                       midpoint = 0, limit = c(-1,1), space = "Lab", 
                       name = NULL) +
  scale_x_discrete(drop = FALSE) +
  scale_y_discrete(drop = FALSE) +
  geom_text(aes(var2, var1, label = value), color = "black", size = 3) +
  labs(x = NULL, y = NULL, title = "SECOM (Subset of Data)") +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 45, vjust = 1, size = 12, hjust = 1),
        axis.text.y = element_text(size = 12),
        plot.title = element_text(hjust = 0.5),
        panel.grid.major = element_blank(),
        axis.ticks = element_blank(),
        legend.position = c(0.3, 0.8),
        legend.direction = "horizontal")+
  guides(fill = guide_colorbar(barwidth = 7, barheight = 1,
                               title.position = "top", title.hjust = 0.5)) +
  coord_fixed()
```

### SECOM (full data)

```{r}
pseqs = list(c(phyloseq1, phyloseq2))
pesudo = 0; zero_cut = 0.5; corr_cut = 0.5; lib_cut = 1000
wins_quant = c(0.05, 0.95); method = "pearson"; soft = FALSE; thresh_len = 20
n_cv = 10; seed = 123; thresh_hard = 0.3; max_p = 0.005; n_cl = 1

res_secom2 = secom_linear(pseqs, pesudo, zero_cut, corr_cut, lib_cut, 
                          wins_quant, method, soft, thresh_len, n_cv, 
                          seed, thresh_hard, max_p, n_cl)

df = res_secom2$corr_fl
df_p = data.frame(get_upper_tri(df)) %>%
  rownames_to_column("var1") %>%
  pivot_longer(cols = colnames(df), names_to = "var2", 
               values_to = "value") %>%
  filter(!is.na(value)) %>%
  mutate(value = round(value, 2))
df_p$var1 = factor(df_p$var1, levels = colnames(df))
df_p$var2 = factor(df_p$var2, levels = colnames(df))

p_secom2 = df_p %>%
  ggplot(aes(var2, var1, fill = value)) +
  geom_tile(color = "black") +
  scale_fill_gradient2(low = "blue", high = "red", mid = "white", na.value = "white",
                       midpoint = 0, limit = c(-1,1), space = "Lab", 
                       name = NULL) +
  scale_x_discrete(drop = FALSE) +
  scale_y_discrete(drop = FALSE) +
  geom_text(aes(var2, var1, label = value), color = "black", size = 3) +
  labs(x = NULL, y = NULL, title = "SECOM (Full Data)") +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 45, vjust = 1, size = 12, hjust = 1),
        axis.text.y = element_text(size = 12),
        plot.title = element_text(hjust = 0.5),
        panel.grid.major = element_blank(),
        axis.ticks = element_blank(),
        legend.position = c(0.3, 0.8),
        legend.direction = "horizontal")+
  guides(fill = guide_colorbar(barwidth = 7, barheight = 1,
                               title.position = "top", title.hjust = 0.5)) +
  coord_fixed()
```

### SparCC (subset of data)

```{r}
O = abundances(subset_taxa(genus_data, Genus %in% topx2))
tax_keep = rownames(O)
df_sparcc1 = sparcc(t(O))$Cor
colnames(df_sparcc1) = rownames(O)
rownames(df_sparcc1) = rownames(O)
df_sparcc1 = hard_thresh(df_sparcc1, th = 0.3)
df = df_sparcc1
df_p = data.frame(get_upper_tri(df)) %>%
  rownames_to_column("var1") %>%
  pivot_longer(cols = colnames(df), names_to = "var2", 
               values_to = "value") %>%
  filter(!is.na(value)) %>%
  mutate(value = round(value, 2))
df_p$var1 = factor(df_p$var1, levels = colnames(df))
df_p$var2 = factor(df_p$var2, levels = colnames(df))

p_sparcc1 = df_p %>%
  ggplot(aes(var2, var1, fill = value)) +
  geom_tile(color = "black") +
  scale_fill_gradient2(low = "blue", high = "red", mid = "white", na.value = "white",
                       midpoint = 0, limit = c(-1,1), space = "Lab", 
                       name = NULL) +
  scale_x_discrete(drop = FALSE) +
  scale_y_discrete(drop = FALSE) +
  geom_text(aes(var2, var1, label = value), color = "black", size = 3) +
  labs(x = NULL, y = NULL, title = "SparCC (Subset of Data)") +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 45, vjust = 1, size = 12, hjust = 1),
        axis.text.y = element_text(size = 12),
        plot.title = element_text(hjust = 0.5),
        panel.grid.major = element_blank(),
        axis.ticks = element_blank(),
        legend.position = c(0.3, 0.8),
        legend.direction = "horizontal")+
  guides(fill = guide_colorbar(barwidth = 7, barheight = 1,
                               title.position = "top", title.hjust = 0.5)) +
  coord_fixed()
```

### SparCC (full data)

```{r, eval=FALSE}
O = abundances(genus_data)
df_sparcc2 = sparcc(t(O))$Cor
colnames(df_sparcc2) = rownames(O)
rownames(df_sparcc2) = rownames(O)
write_csv(as.data.frame(df_sparcc2), "../outputs/others/df_sparcc2.csv")
```

```{r}
df_sparcc2 = read_csv("../outputs/others/df_sparcc2.csv")
rownames(df_sparcc2) = colnames(df_sparcc2)
df_sparcc2 = as.matrix(df_sparcc2)[tax_keep, tax_keep]
df_sparcc2 = hard_thresh(df_sparcc2, th = 0.3)
df = df_sparcc2
df_p = data.frame(get_upper_tri(df)) %>%
  rownames_to_column("var1") %>%
  pivot_longer(cols = colnames(df), names_to = "var2", 
               values_to = "value") %>%
  filter(!is.na(value)) %>%
  mutate(value = round(value, 2))
df_p$var1 = factor(df_p$var1, levels = colnames(df))
df_p$var2 = factor(df_p$var2, levels = colnames(df))

p_sparcc2 = df_p %>%
  ggplot(aes(var2, var1, fill = value)) +
  geom_tile(color = "black") +
  scale_fill_gradient2(low = "blue", high = "red", mid = "white", na.value = "white",
                       midpoint = 0, limit = c(-1,1), space = "Lab", 
                       name = NULL) +
  scale_x_discrete(drop = FALSE) +
  scale_y_discrete(drop = FALSE) +
  geom_text(aes(var2, var1, label = value), color = "black", size = 3) +
  labs(x = NULL, y = NULL, title = "SparCC (Full Data)") +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 45, vjust = 1, size = 12, hjust = 1),
        axis.text.y = element_text(size = 12),
        plot.title = element_text(hjust = 0.5),
        panel.grid.major = element_blank(),
        axis.ticks = element_blank(),
        legend.position = c(0.3, 0.8),
        legend.direction = "horizontal")+
  guides(fill = guide_colorbar(barwidth = 7, barheight = 1,
                               title.position = "top", title.hjust = 0.5)) +
  coord_fixed()
```

### Output

```{r, fig.width=10, fig.height=10}
p_seq_depth = ggarrange(p_secom1, p_secom2, p_sparcc1, p_sparcc2, 
                        labels = c("a", "b", "c", "d"), 
                        nrow = 2, ncol = 2, common.legend = TRUE, legend = "bottom")
p_seq_depth
ggsave(plot = p_seq_depth, "../images/supp/supp_seq_depth.pdf", height = 10, width = 12)   
ggsave(plot = p_seq_depth, "../images/supp/supp_seq_depth.jpeg", height = 10, width = 12, dpi = 300)
```

# 4. Original data vs. rank-transformed data

```{r, fig.height=8, fig.width=10}
set.seed(123)

# Linear
df = data.frame(x = runif(100, min = -2, max = 2)) %>% 
  mutate(y = 2 * x + rnorm(100))
rank_df = df %>% 
  transmute(x = rank(x), y = rank(y))
p_df = bind_rows(df, rank_df) %>% 
  mutate(group = rep(c("Original", "Rank"), each = 100))
ann_df = data.frame(x = c(0, 50), y = c(5, 98), lab = "Text",
                    group = c("Original", "Rank"),
                    label = c(paste0("dCor = ", round(dcor(x = df$x, y = df$y), 2)),
                              paste0("dCor = ", round(dcor(x = rank_df$x, y = rank_df$y), 2))))
p1 = p_df %>% 
  ggplot(aes(x = x, y = y)) +
  geom_point() + 
  facet_wrap(.~group, scales = "free") +
  theme_bw() +
  geom_label(data = ann_df, mapping = aes(x = x, y = y, group = group, label = label),
             color = "orange", size = 4, fontface = "bold") +
  theme(strip.background = element_rect(fill = "white"))

# Nonlinear
df = data.frame(x = runif(100, min = -2, max = 2)) %>% 
  mutate(y = x^2 + rnorm(100))
rank_df = df %>% 
  transmute(x = rank(x), y = rank(y))
p_df = bind_rows(df, rank_df) %>% 
  mutate(group = rep(c("Original", "Rank"), each = 100))
ann_df = data.frame(x = c(0, 50), y = c(5, 95), lab = "Text",
                    group = c("Original", "Rank"),
                    label = c(paste0("dCor = ", round(dcor(x = df$x, y = df$y), 2)),
                              paste0("dCor = ", round(dcor(x = rank_df$x, y = rank_df$y), 2))))
p2 = p_df %>% 
  ggplot(aes(x = x, y = y)) +
  geom_point() + 
  facet_wrap(.~group, scales = "free") +
  theme_bw() +
  geom_label(data = ann_df, mapping = aes(x = x, y = y, group = group, label = label),
             color = "orange", size = 4, fontface = "bold") +
  theme(strip.background = element_rect(fill = "white"))

p = ggarrange(p1, p2, nrow = 2, ncol = 1, labels = c("a", "b"))
print(p)
ggsave(plot = p, "../images/supp/supp_org_rank.pdf", height = 8, width = 10)   
ggsave(plot = p, "../images/supp/supp_org_rank.jpeg", height = 8, width = 10, dpi = 300)
```

# Session information

```{r, message = FALSE, warning = FALSE, comment = NA}
sessionInfo()
```









